subsys/net/lib/tls_credentials/tls_credentials_shell.c

```diff
@@ -0,0 +1,813 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(tls_credentials_shell, CONFIG_TLS_CREDENTIALS_LOG_LEVEL);
+
+#include <zephyr/kernel.h>
+#include <zephyr/shell/shell.h>
+#include <zephyr/sys/base64.h>
+#include <zephyr/net/tls_credentials.h>
+#include "tls_internal.h"
+#include <strings.h>
+#include <ctype.h>
+
+//TODO: fix newlines
+
+enum cred_storage_fmt {
+	/* Credential is stored as a string and will be passed between the shell and storage
+	 * unmodified.
+	 */
+	CRED_STORAGE_FMT_STRING,
+
+	/* Credential is stored as raw binary, and is parsed from base64 before storage and encoded
+	 * back into base64 when retrieved via the shell.
+	 */
+	CRED_STORAGE_FMT_BINARY,
+};
+
+struct cred_type_string {
+	char* name;
+	enum tls_credential_type type;
+};
+
+/* The first entry in each credential type group will be used for human-readable shell
+ * output. The last will be used for compact shell output. The rest are accepted synonyms.
+ */
+struct cred_type_string type_strings[] = {
+	{"CA_CERT",		TLS_CREDENTIAL_CA_CERTIFICATE},
+	{"CA",			TLS_CREDENTIAL_CA_CERTIFICATE},
+
+	{"SERVER_CERT", 	TLS_CREDENTIAL_SERVER_CERTIFICATE},
+	{"SELF_CERT",		TLS_CREDENTIAL_SERVER_CERTIFICATE},
+	{"SELF",		TLS_CREDENTIAL_SERVER_CERTIFICATE},
+	{"SERV",		TLS_CREDENTIAL_SERVER_CERTIFICATE},
+
+	{"PRIVATE_KEY",		TLS_CREDENTIAL_PRIVATE_KEY},
+	{"PK",			TLS_CREDENTIAL_PRIVATE_KEY},
+
+	{"PRE_SHARED_KEY",	TLS_CREDENTIAL_PSK},
+	{"PSK",			TLS_CREDENTIAL_PSK},
+
+	{"PRE_SHARED_KEY_ID",	TLS_CREDENTIAL_PSK_ID},
+	{"PSK_ID",		TLS_CREDENTIAL_PSK_ID}
+};
+
+#define ANY_KEYWORD "any"
+
+#define NUM_TYPE_STRINGS (sizeof(type_strings)/sizeof(type_strings[0]))
+
+/* This is so that we can output base64 in chunks of this length if necessary */
+BUILD_ASSERT(
+	(CONFIG_TLS_CREDENTIALS_SHELL_CRED_OUTPUT_WIDTH % 4) == 0,
+	"CONFIG_TLS_CREDENTIALS_SHELL_CRED_OUTPUT_WIDTH must be a multiple of 4."
+);
+
+/* Output buffers used for printing credentials and digests.
+ * One extra byte included for NULL termination.
+ */
+static char cred_out_buf[CONFIG_TLS_CREDENTIALS_SHELL_CRED_OUTPUT_WIDTH + 1];
+static char cred_digest_buf[CONFIG_TLS_CREDENTIALS_SHELL_DIGEST_BUF_SIZE + 1];
+
+/* Internal buffer used for storing and retrieving credentials.
+ * +1 byte for potential NULL termination.
+ */
+static char cred_buf[CONFIG_TLS_CREDENTIALS_SHELL_CRED_BUF_SIZE + 1];
+static size_t cred_written = 0;
+
+/* Some backends (namely, the volatile backend) store a reference rather than a copy of passed-in
+ * credentials. For these backends, we need to copy incoming credentials onto the heap before
+ * attempting to store them.
+ *
+ * Since the backend in use is determind at build time by KConfig, so is this behavior.
+ * If multi/dynamic-backend support is ever added, this will need to be updated.
+ */
+#define COPY_CREDENTIALS_TO_HEAP IS_ENABLED(CONFIG_TLS_CREDENTIALS_BACKEND_VOLATILE)
```

Comment:
I suggest to do this other way around, so just have
`#define COPY_CREDENTIALS_TO_HEAP CONFIG_TLS_CREDENTIALS_BACKEND_VOLATILE`
or something even shorter define.
And then when you use it below
`if (IS_ENABLED(COPY_CREDENTIALS_TO_HEAP)) {`
this way it is clear what we are checking and what is expected from the check. Hiding the `IS_ENABLED()` is not needed.